### DISCUSSION TITLE
JsonMap 字段如何查询

### DISCUSSION THREAD

**Discussion by woodcoal** [dotnetcore/FreeSql#1152] [Q&A] [answered]
JsonMap 字段如果要查询的话，是使用字符串包含呢还是使用 map 的对象？
个人理解貌似直接查询的话 map 的对象可能不行。
但是使用表达式的话，对象又不能用 Contains

**Comment by 2881099** [comment]
see 自定义表达式树函数解析 
https://github.com/dotnetcore/FreeSql/wiki/%e8%a1%a8%e8%be%be%e5%bc%8f%e5%87%bd%e6%95%b0#%E8%87%AA%E5%AE%9A%E4%B9%89%E8%A7%A3%E6%9E%90

**Comment by woodcoal** [comment]
我的理解是 JsonMap 使用的是文本字段保存，而如果要查找映射对象中的值，是不是就存在一个序列化的过程。

比如：
```
class ABC {
    [Column(IsIdentity = true)]
    public int Id { get; set; }
    public string Title { get; set; }

    [JsonMap]
    public DEF Info { get; set; }
}

class DEF {
    public int Index { get; set; }
    public string Name { get; set; }
}
```

我要查询 ABC 中 Info 字段 Index = 100 的数据
```
fsql.Select<ABC>()
  .Where(a => a.Info.Index == 100)
  .ToList();
```
这样操作可行？

像 mysql， sql server 这些原生是不支持 JSON 的恐怕这样查询不行吧。
除非二次查询，一次将所有数据都加载，反序列化出对象，然后再从所有对象中取匹配。

**Comment by 2881099** [comment, marked as answer]
@woodcoal 是的，新版本 mysql/sqlserver 已经支持 json 类型了，有对应的 SQL 函数可以达到该功能

比如 postgresql 数据库 a.Info.Index == 100 翻译成 SQL 是 a."info"->>"index" = 100

所有数据库中 postgresql 对 json 类型支持最好。

看下这个内容：PostgreSQL pgsql jsonb 类型 POCO 访问扩展现实方案 https://github.com/dotnetcore/FreeSql/discussions/1071

**Reply by woodcoal** [reply]
谢谢
